internal/provider/common_key.go

```diff
@@ -0,0 +1,190 @@
+package provider
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/ed25519"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"golang.org/x/crypto/ssh"
+)
+
+// keyGenerator extracts data from the given *schema.ResourceData,
+// and generates a new public/private key-pair according to the
+// selected algorithm.
+type keyGenerator func(d *schema.ResourceData) (crypto.PrivateKey, error)
+
+// keyParser parses a private key from the given []byte,
+// according to the selected algorithm.
+type keyParser func([]byte) (crypto.PrivateKey, error)
+
+// keyGenerators provides a keyGenerator given a specific Algorithm.
+var keyGenerators = map[Algorithm]keyGenerator{
+	RSA: func(d *schema.ResourceData) (crypto.PrivateKey, error) {
+		rsaBits := d.Get("rsa_bits").(int)
+		return rsa.GenerateKey(rand.Reader, rsaBits)
+	},
+	ECDSA: func(d *schema.ResourceData) (crypto.PrivateKey, error) {
+		curve := ECDSACurve(d.Get("ecdsa_curve").(string))
+		switch curve {
+		case P224:
+			return ecdsa.GenerateKey(elliptic.P224(), rand.Reader)
+		case P256:
+			return ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+		case P384:
+			return ecdsa.GenerateKey(elliptic.P384(), rand.Reader)
+		case P521:
+			return ecdsa.GenerateKey(elliptic.P521(), rand.Reader)
+		default:
+			return nil, fmt.Errorf("invalid ECDSA curve; supported values are: %v", SupportedECDSACurves())
+		}
+	},
+	ED25519: func(d *schema.ResourceData) (crypto.PrivateKey, error) {
+		_, key, err := ed25519.GenerateKey(rand.Reader)
+		if err != nil {
+			return nil, fmt.Errorf("failed to generate ED25519 key: %s", err)
+		}
+		return &key, err
+	},
+}
+
+// keyParsers provides a keyParser given a specific Algorithm.
+var keyParsers = map[Algorithm]keyParser{
+	RSA: func(der []byte) (crypto.PrivateKey, error) {
+		return x509.ParsePKCS1PrivateKey(der)
+	},
+	ECDSA: func(der []byte) (crypto.PrivateKey, error) {
+		return x509.ParseECPrivateKey(der)
+	},
+	ED25519: func(der []byte) (crypto.PrivateKey, error) {
+		return x509.ParsePKCS8PrivateKey(der)
+	},
+}
+
+// parsePrivateKeyPEM takes a slide of bytes containing a private key
+// encoded in [PEM (RFC 1421)](https://datatracker.ietf.org/doc/html/rfc1421) format,
+// and returns a crypto.PrivateKey implementation, together with the Algorithm used by the key.
+func parsePrivateKeyPEM(keyPEMBytes []byte) (crypto.PrivateKey, Algorithm, error) {
+	pemBlock, rest := pem.Decode(keyPEMBytes)
+	if pemBlock == nil {
+		return nil, "", fmt.Errorf("failed to decode PEM block: decoded bytes %d, undecoded %d", len(keyPEMBytes)-len(rest), len(rest))
+	}
+
+	// Map PEM Preamble of the Private Key to the corresponding Algorithm
+	algorithm, err := PEMPreamblePrivateKey(pemBlock.Type).Algorithm()
```

Comment:
OK so, I have dug into the code of `crypto/ssh` and `crypto.x509` and the Preamble (i.e. the `pem.Block.Type` field) are indeed what drives the parsing decision in all the "switches".

Given the ambiguity in the RFC that I highlighted in our private chat, I feel like the best course of action is to imitate what Golang libraries already do.

I will of course integrate any feedback I might get back from the security team, so far nothing.

Additional to all of this, by poking me on this it made me realise that there was a corner case currently unhandled: the Preamble `BEGING PRIVATE KEY` can be used for more than just ED25519, but it can instead be used for RSA and ECDSA too.

Here is the signature of the method designated to parse it:

```golang
// ParsePKCS8PrivateKey parses an unencrypted private key in PKCS #8, ASN.1 DER form.
//
// It returns a *rsa.PrivateKey, a *ecdsa.PrivateKey, or a ed25519.PrivateKey.
// More types might be supported in the future.
//
// This kind of key is commonly encoded in PEM blocks of type "PRIVATE KEY".
func ParsePKCS8PrivateKey(der []byte) (key interface{}, err error) {
  // ...
}
```

So, I have restructured things to handle the key identification AFTER the key parsing.